server/src/main/java/org/elasticsearch/index/mapper/MappingParser.java

```diff
@@ -124,7 +124,7 @@ Mapping parse(@Nullable String type, MergeReason reason, Map<String, Object> map
         Map<Class<? extends MetadataFieldMapper>, MetadataFieldMapper> metadataMappers = metadataMappersSupplier.get();
         Map<String, Object> meta = null;
 
-        boolean isSourceSynthetic = mappingParserContext.getIndexSettings().getMode().isSyntheticSourceEnabled();
+        boolean isSourceSynthetic = SourceFieldMapper.isSynthetic(mappingParserContext.getIndexSettings());
```

Comment:
This is an interesting place. At some point this will just be equal to the index setting value but for now it is possible that `_source.mode` mapping parameters overrides it so we keep existing logic below that overrides this with `mode` from the mapper.